index.bs

```diff
@@ -10496,6 +10504,40 @@ that does specify [{{SecureContext}}].
     </pre>
 </div>
 
+<h4 id="StringContext" extended-attribute lt="StringContext">[StringContext]</h4>
+
+If the [{{StringContext}}] [=extended attribute=] appears on {{DOMString}} or {{USVString}}, it
+modifies how the value is converted to the IDL type, causing additional value validation to
+adhere to the context the string is used in.
+
+The [{{StringContext}}] extended attribute must [=takes an identifier|take an identifier=]. The [=identifier=]
+must be one of "<code>html</code>", "<code>script-url</code>" and "<code>script</code>".
+
+A type annotated with the [{{StringContext}}] extended attribute must not appear in a construct
+that is not a [=regular attribute=] or a [=regular operation=]. A type annotated with the [{{StringContext}}]
```

Comment:
OK, reading this more carefully, I don't understand what this means.  Is `void foo(sequence<[StringContext=html] DOMString>);` ok?  Why or why not?  What does "construct" even mean here?

What about a record?  A dictionary argument?  A union?  Is a dictionary a "separate construct" while a record is not, just because a dictionary has a separate IDL definition while a record or union is inferred?  Or are they both "separate constructs"?  But my impression was that you wanted to allow this new thing inside unions, which are just as "separate construct" as records.  Or is that out of date and the intent is to limit this thing to types of attributes and operation arguments?  If so, that seems like the right thing to say explicitly.

Reply:
It's the latter. I'll fix the language.

Reply:
OK.  So no more unions involved; this mechanism is replacing that?

If so, then the right place to put this might be not inside generic DOMString/USVString conversion but specifically in operation argument and attribute setter type conversions.  That would avoid the ambient "related construct" bit and make it much clearer when these things actually have an effect.

Reply:
Though again, at that point it could in theory also just happen at the top of the operation steps instead of intertwined with the other argument conversions, in which case it would not need any IDL hooks at all.  It's not clear to me which is easier or harder to reason about and implement in practice, or if there's a practical difference between them. 

All this assuming I now understand the proposal correctly...

Reply:
> OK. So no more unions involved; this mechanism is replacing that?

Yes, the extended attribute is to avoid using type unions in most of the places.

> If so, then the right place to put this might be [...] in operation argument and attribute setter type conversions. 
So something like https://w3c.github.io/webappsec-trusted-types/dist/spec/#!trustedtypes-extended-attribute ? That's the previous attempt.

Reply:
I updated the PR such that "the intent is to limit this thing to types of attributes and operation arguments" is explicit.

Reply:
> So something like

In terms of my options in https://github.com/heycam/webidl/pull/841#issuecomment-586354298 this is option 3 or 4, I think, right?  I was talking about option 2 here.

Reply:
Resolving, since we chose the option 1 here in the end.